File: enrich/compliance.go
package enrich

import (
	"strconv"
	"strings"
)

// ComplianceResult evaluates a TLS configuration against a compliance framework.
type ComplianceResult struct {
	Framework        string              `json:"framework"`
	DisplayName      string              `json:"display_name"`
	MeetsRequirements bool               `json:"meets_requirements"`
	Findings         []ComplianceFinding `json:"findings"`
}

// ComplianceFinding is a single pass/fail/warn check within a framework.
type ComplianceFinding struct {
	Requirement string `json:"requirement"`
	Status      string `json:"status"` // "pass", "fail", "warn"
	Detail      string `json:"detail"`
}

// ComplianceInput provides the data needed for compliance evaluation.
type ComplianceInput struct {
	Protocols      []string
	Ciphers        []CipherForCompliance
	KeyAlg         string
	KeySize        int
	ForwardSecrecy bool
	ChainValid     bool
	DaysRemaining  int
	OCSPStapling   bool
	HSTSEnabled    bool
}

// CipherForCompliance is a simplified cipher reference for compliance checks.
type CipherForCompliance struct {
	Name     string
	Strength string
}

// EvaluateCompliance runs all compliance frameworks against the input.
func EvaluateCompliance(data ComplianceInput) []ComplianceResult {
	return []ComplianceResult{
		evaluatePCI(data),
		evaluateNIST(data),
		evaluateHIPAA(data),
	}
}

// ─── Shared helpers ─────────────────────────────────────────────────

func hasLegacyProtocol(protocols []string) bool {
	for _, p := range protocols {
		if p == "TLS 1.0" || p == "TLS 1.1" || p == "SSLv3" || p == "SSLv2" {
			return true
		}
	}
	return false
}

func hasTLS12Plus(protocols []string) bool {
	for _, p := range protocols {
		if p == "TLS 1.2" || p == "TLS 1.3" {
			return true
		}
	}
	return false
}

var weakCipherPatterns = []string{"RC4", "DES", "3DES", "NULL", "EXPORT", "anon", "MD5"}

func findWeakCiphers(ciphers []CipherForCompliance) []string {
	var weak []string
	for _, c := range ciphers {
		if c.Strength == "insecure" || c.Strength == "weak" {
			weak = append(weak, c.Name)
			continue
		}
		for _, pat := range weakCipherPatterns {
			if strings.Contains(strings.ToUpper(c.Name), strings.ToUpper(pat)) {
				weak = append(weak, c.Name)
				break
			}
		}
	}
	return weak
}

func formatWeakList(weak []string) string {
	if len(weak) <= 3 {
		return strings.Join(weak, ", ")
	}
	return strings.Join(weak[:3], ", ") + " +" + strconv.Itoa(len(weak)-3) + " more"
}

// ─── PCI DSS 4.0 ───────────────────────────────────────────────────

func evaluatePCI(data ComplianceInput) ComplianceResult {
	var findings []ComplianceFinding

	if !hasTLS12Plus(data.Protocols) {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "fail", "TLS 1.2 or higher not supported"})
	} else {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "pass", "TLS 1.2+ supported"})
	}

	if hasLegacyProtocol(data.Protocols) {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "fail", "Legacy protocols enabled"})
	}

	weak := findWeakCiphers(data.Ciphers)
	if len(weak) > 0 {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "fail", "Weak/insecure ciphers: " + formatWeakList(weak)})
	} else {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "pass", "No weak or insecure ciphers"})
	}

	if !data.ForwardSecrecy {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "fail", "Forward secrecy not supported"})
	} else {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "pass", "Forward secrecy enabled"})
	}

	if !data.ChainValid || data.DaysRemaining <= 0 {
		detail := "Invalid certificate chain"
		if data.DaysRemaining <= 0 {
			detail = "Certificate expired"
		}
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "fail", detail})
	} else {
		findings = append(findings, ComplianceFinding{"Req 4.2.1", "pass", "Valid certificate chain"})
	}

	if !data.HSTSEnabled {
		findings = append(findings, ComplianceFinding{"Best Practice", "warn", "HSTS not enabled"})
	} else {
		findings = append(findings, ComplianceFinding{"Best Practice", "pass", "HSTS enabled"})
	}

	meets := true
	for _, f := range findings {
		if f.Status == "fail" {
			meets = false
			break
		}
	}

	return ComplianceResult{"pci-dss-4", "PCI DSS 4.0", meets, findings}
}

// ─── NIST SP 800-52r2 ───────────────────────────────────────────────

func evaluateNIST(data ComplianceInput) ComplianceResult {
	var findings []ComplianceFinding

	if !hasTLS12Plus(data.Protocols) {
		findings = append(findings, ComplianceFinding{"Sec 3.1", "fail", "TLS 1.2 or higher not supported"})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 3.1", "pass", "TLS 1.2+ supported"})
	}

	hasTLS13 := false
	for _, p := range data.Protocols {
		if p == "TLS 1.3" {
			hasTLS13 = true
		}
	}
	if !hasTLS13 {
		findings = append(findings, ComplianceFinding{"Sec 3.1", "warn", "TLS 1.3 not supported (recommended)"})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 3.1", "pass", "TLS 1.3 supported"})
	}

	if hasLegacyProtocol(data.Protocols) {
		findings = append(findings, ComplianceFinding{"Sec 3.1", "fail", "Legacy protocols enabled"})
	}

	// Key size
	algUpper := strings.ToUpper(data.KeyAlg)
	if strings.Contains(algUpper, "RSA") {
		if data.KeySize > 0 && data.KeySize < 2048 {
			findings = append(findings, ComplianceFinding{"Sec 3.3", "fail",
				"RSA key " + strconv.Itoa(data.KeySize) + "-bit < 2048-bit minimum"})
		} else if data.KeySize >= 2048 {
			findings = append(findings, ComplianceFinding{"Sec 3.3", "pass",
				"RSA " + strconv.Itoa(data.KeySize) + "-bit key meets minimum"})
		}
	} else if strings.Contains(algUpper, "ECDSA") || strings.Contains(algUpper, "EC") {
		if data.KeySize > 0 && data.KeySize < 256 {
			findings = append(findings, ComplianceFinding{"Sec 3.3", "fail",
				"ECDSA key " + strconv.Itoa(data.KeySize) + "-bit < 256-bit minimum"})
		} else if data.KeySize >= 256 {
			findings = append(findings, ComplianceFinding{"Sec 3.3", "pass",
				"ECDSA " + strconv.Itoa(data.KeySize) + "-bit key meets minimum"})
		}
	}

	weak := findWeakCiphers(data.Ciphers)
	if len(weak) > 0 {
		findings = append(findings, ComplianceFinding{"Sec 3.3", "fail", "Weak/insecure ciphers: " + formatWeakList(weak)})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 3.3", "pass", "No weak or insecure ciphers"})
	}

	if !data.ForwardSecrecy {
		findings = append(findings, ComplianceFinding{"Sec 3.3", "fail", "Forward secrecy not supported"})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 3.3", "pass", "Forward secrecy enabled"})
	}

	if !data.OCSPStapling {
		findings = append(findings, ComplianceFinding{"Sec 4.4", "warn", "OCSP stapling not present (recommended)"})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 4.4", "pass", "OCSP stapling enabled"})
	}

	if !data.ChainValid || data.DaysRemaining <= 0 {
		detail := "Invalid certificate chain"
		if data.DaysRemaining <= 0 {
			detail = "Certificate expired"
		}
		findings = append(findings, ComplianceFinding{"Sec 4", "fail", detail})
	} else {
		findings = append(findings, ComplianceFinding{"Sec 4", "pass", "Valid certificate chain"})
	}

	meets := true
	for _, f := range findings {
		if f.Status == "fail" {
			meets = false
			break
		}
	}

	return ComplianceResult{"nist-800-52r2", "NIST 800-52r2", meets, findings}
}

// ─── HIPAA ──────────────────────────────────────────────────────────

func evaluateHIPAA(data ComplianceInput) ComplianceResult {
	var findings []ComplianceFinding

	if !hasTLS12Plus(data.Protocols) {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "fail", "TLS 1.2 or higher not supported"})
	} else {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "pass", "TLS 1.2+ supported"})
	}

	if hasLegacyProtocol(data.Protocols) {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "fail", "Legacy protocols enabled"})
	}

	weak := findWeakCiphers(data.Ciphers)
	if len(weak) > 0 {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "fail", "Insecure ciphers: " + formatWeakList(weak)})
	} else {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "pass", "Strong encryption (AES) in use"})
	}

	if !data.ChainValid || data.DaysRemaining <= 0 {
		detail := "Invalid certificate chain"
		if data.DaysRemaining <= 0 {
			detail = "Certificate expired"
		}
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "fail", detail})
	} else {
		findings = append(findings, ComplianceFinding{"§164.312(e)(1)", "pass", "Valid certificate chain"})
	}

	if !data.ForwardSecrecy {
		findings = append(findings, ComplianceFinding{"Best Practice", "warn", "Forward secrecy not supported (recommended)"})
	} else {
		findings = append(findings, ComplianceFinding{"Best Practice", "pass", "Forward secrecy enabled"})
	}

	meets := true
	for _, f := range findings {
		if f.Status == "fail" {
			meets = false
			break
		}
	}

	return ComplianceResult{"hipaa", "HIPAA", meets, findings}
}

File: src/compliance.ts
// ─── Compliance Mapping ─────────────────────────────────────────────
// Evaluates TLS scan data against transport encryption requirements
// for PCI DSS 4.0, NIST SP 800-52r2, and HIPAA.
//
// Important: these are transport-layer checks only. Passing here means
// "meets transport encryption requirements for <framework>" — NOT full
// compliance with the framework itself.

export interface ComplianceResult {
  framework: string;
  display_name: string;
  meets_requirements: boolean;
  findings: ComplianceFinding[];
}

export interface ComplianceFinding {
  requirement: string;
  status: 'pass' | 'fail' | 'warn';
  detail: string;
}

export interface ComplianceInput {
  protocols: string[];
  ciphers: Array<{ name: string; strength: string }>;
  cipher_summary: { strong: number; acceptable: number; weak: number; insecure: number };
  key_alg: string;
  key_size: number;
  forward_secrecy: boolean;
  chain_valid: boolean;
  days_remaining: number;
  ocsp_stapling: boolean;
  hsts?: { enabled: boolean } | null;
}

// ─── Shared checks ──────────────────────────────────────────────────

const WEAK_CIPHER_PATTERNS = [
  /RC4/i, /DES/i, /3DES/i, /NULL/i, /EXPORT/i, /anon/i, /MD5/i,
];

function hasLegacyProtocol(protocols: string[]): boolean {
  return protocols.some(p => p === 'TLS 1.0' || p === 'TLS 1.1' || p === 'SSLv3' || p === 'SSLv2');
}

function hasTLS12Plus(protocols: string[]): boolean {
  return protocols.some(p => p === 'TLS 1.2' || p === 'TLS 1.3');
}

function hasWeakCiphers(ciphers: Array<{ name: string; strength: string }>): string[] {
  const weak: string[] = [];
  for (const c of ciphers) {
    if (c.strength === 'insecure' || c.strength === 'weak') {
      weak.push(c.name);
      continue;
    }
    for (const pat of WEAK_CIPHER_PATTERNS) {
      if (pat.test(c.name)) {
        weak.push(c.name);
        break;
      }
    }
  }
  return weak;
}

// ─── PCI DSS 4.0 (Requirement 4.2.1) ───────────────────────────────

function evaluatePCI(data: ComplianceInput): ComplianceResult {
  const findings: ComplianceFinding[] = [];

  // TLS 1.2+ required
  if (!hasTLS12Plus(data.protocols)) {
    findings.push({ requirement: 'Req 4.2.1', status: 'fail', detail: 'TLS 1.2 or higher not supported' });
  } else {
    findings.push({ requirement: 'Req 4.2.1', status: 'pass', detail: 'TLS 1.2+ supported' });
  }

  // No legacy protocols
  if (hasLegacyProtocol(data.protocols)) {
    const legacy = data.protocols.filter(p => p === 'TLS 1.0' || p === 'TLS 1.1' || p === 'SSLv3' || p === 'SSLv2');
    findings.push({ requirement: 'Req 4.2.1', status: 'fail', detail: `Legacy protocols enabled: ${legacy.join(', ')}` });
  }

  // No weak ciphers
  const weak = hasWeakCiphers(data.ciphers);
  if (weak.length > 0) {
    findings.push({ requirement: 'Req 4.2.1', status: 'fail', detail: `Weak/insecure ciphers: ${weak.slice(0, 3).join(', ')}${weak.length > 3 ? ` +${weak.length - 3} more` : ''}` });
  } else {
    findings.push({ requirement: 'Req 4.2.1', status: 'pass', detail: 'No weak or insecure ciphers' });
  }

  // Forward secrecy
  if (!data.forward_secrecy) {
    findings.push({ requirement: 'Req 4.2.1', status: 'fail', detail: 'Forward secrecy not supported' });
  } else {
    findings.push({ requirement: 'Req 4.2.1', status: 'pass', detail: 'Forward secrecy enabled' });
  }

  // Valid certificate
  if (!data.chain_valid || data.days_remaining <= 0) {
    findings.push({ requirement: 'Req 4.2.1', status: 'fail', detail: data.days_remaining <= 0 ? 'Certificate expired' : 'Invalid certificate chain' });
  } else {
    findings.push({ requirement: 'Req 4.2.1', status: 'pass', detail: 'Valid certificate chain' });
  }

  // HSTS (recommended, not required by spec)
  if (!data.hsts?.enabled) {
    findings.push({ requirement: 'Best Practice', status: 'warn', detail: 'HSTS not enabled' });
  } else {
    findings.push({ requirement: 'Best Practice', status: 'pass', detail: 'HSTS enabled' });
  }

  return {
    framework: 'pci-dss-4',
    display_name: 'PCI DSS 4.0',
    meets_requirements: findings.every(f => f.status !== 'fail'),
    findings,
  };
}

// ─── NIST SP 800-52r2 ───────────────────────────────────────────────

function evaluateNIST(data: ComplianceInput): ComplianceResult {
  const findings: ComplianceFinding[] = [];

  // TLS 1.2+ required
  if (!hasTLS12Plus(data.protocols)) {
    findings.push({ requirement: 'Sec 3.1', status: 'fail', detail: 'TLS 1.2 or higher not supported' });
  } else {
    findings.push({ requirement: 'Sec 3.1', status: 'pass', detail: 'TLS 1.2+ supported' });
  }

  // TLS 1.3 preferred
  if (!data.protocols.includes('TLS 1.3')) {
    findings.push({ requirement: 'Sec 3.1', status: 'warn', detail: 'TLS 1.3 not supported (recommended)' });
  } else {
    findings.push({ requirement: 'Sec 3.1', status: 'pass', detail: 'TLS 1.3 supported' });
  }

  // No legacy protocols
  if (hasLegacyProtocol(data.protocols)) {
    const legacy = data.protocols.filter(p => p === 'TLS 1.0' || p === 'TLS 1.1' || p === 'SSLv3' || p === 'SSLv2');
    findings.push({ requirement: 'Sec 3.1', status: 'fail', detail: `Legacy protocols enabled: ${legacy.join(', ')}` });
  }

  // Key size requirements
  if (data.key_alg.toUpperCase().includes('RSA')) {
    if (data.key_size > 0 && data.key_size < 2048) {
      findings.push({ requirement: 'Sec 3.3', status: 'fail', detail: `RSA key ${data.key_size}-bit < 2048-bit minimum` });
    } else if (data.key_size >= 2048) {
      findings.push({ requirement: 'Sec 3.3', status: 'pass', detail: `RSA ${data.key_size}-bit key meets minimum` });
    }
  } else if (data.key_alg.toUpperCase().includes('ECDSA') || data.key_alg.toUpperCase().includes('EC')) {
    if (data.key_size > 0 && data.key_size < 256) {
      findings.push({ requirement: 'Sec 3.3', status: 'fail', detail: `ECDSA key ${data.key_size}-bit < 256-bit minimum` });
    } else if (data.key_size >= 256) {
      findings.push({ requirement: 'Sec 3.3', status: 'pass', detail: `ECDSA ${data.key_size}-bit key meets minimum` });
    }
  }

  // No weak ciphers
  const weak = hasWeakCiphers(data.ciphers);
  if (weak.length > 0) {
    findings.push({ requirement: 'Sec 3.3', status: 'fail', detail: `Weak/insecure ciphers: ${weak.slice(0, 3).join(', ')}${weak.length > 3 ? ` +${weak.length - 3} more` : ''}` });
  } else {
    findings.push({ requirement: 'Sec 3.3', status: 'pass', detail: 'No weak or insecure ciphers' });
  }

  // Forward secrecy
  if (!data.forward_secrecy) {
    findings.push({ requirement: 'Sec 3.3', status: 'fail', detail: 'Forward secrecy not supported' });
  } else {
    findings.push({ requirement: 'Sec 3.3', status: 'pass', detail: 'Forward secrecy enabled' });
  }

  // OCSP stapling (recommended)
  if (!data.ocsp_stapling) {
    findings.push({ requirement: 'Sec 4.4', status: 'warn', detail: 'OCSP stapling not present (recommended)' });
  } else {
    findings.push({ requirement: 'Sec 4.4', status: 'pass', detail: 'OCSP stapling enabled' });
  }

  // Valid certificate chain
  if (!data.chain_valid || data.days_remaining <= 0) {
    findings.push({ requirement: 'Sec 4', status: 'fail', detail: data.days_remaining <= 0 ? 'Certificate expired' : 'Invalid certificate chain' });
  } else {
    findings.push({ requirement: 'Sec 4', status: 'pass', detail: 'Valid certificate chain' });
  }

  return {
    framework: 'nist-800-52r2',
    display_name: 'NIST 800-52r2',
    meets_requirements: findings.every(f => f.status !== 'fail'),
    findings,
  };
}

// ─── HIPAA (45 CFR §164.312(e)(1)) ─────────────────────────────────

function evaluateHIPAA(data: ComplianceInput): ComplianceResult {
  const findings: ComplianceFinding[] = [];

  // TLS 1.2+ required
  if (!hasTLS12Plus(data.protocols)) {
    findings.push({ requirement: '§164.312(e)(1)', status: 'fail', detail: 'TLS 1.2 or higher not supported' });
  } else {
    findings.push({ requirement: '§164.312(e)(1)', status: 'pass', detail: 'TLS 1.2+ supported' });
  }

  // No legacy protocols
  if (hasLegacyProtocol(data.protocols)) {
    const legacy = data.protocols.filter(p => p === 'TLS 1.0' || p === 'TLS 1.1' || p === 'SSLv3' || p === 'SSLv2');
    findings.push({ requirement: '§164.312(e)(1)', status: 'fail', detail: `Legacy protocols enabled: ${legacy.join(', ')}` });
  }

  // AES required, no weak ciphers
  const weak = hasWeakCiphers(data.ciphers);
  if (weak.length > 0) {
    findings.push({ requirement: '§164.312(e)(1)', status: 'fail', detail: `Insecure ciphers: ${weak.slice(0, 3).join(', ')}${weak.length > 3 ? ` +${weak.length - 3} more` : ''}` });
  } else {
    findings.push({ requirement: '§164.312(e)(1)', status: 'pass', detail: 'Strong encryption (AES) in use' });
  }

  // Valid certificate
  if (!data.chain_valid || data.days_remaining <= 0) {
    findings.push({ requirement: '§164.312(e)(1)', status: 'fail', detail: data.days_remaining <= 0 ? 'Certificate expired' : 'Invalid certificate chain' });
  } else {
    findings.push({ requirement: '§164.312(e)(1)', status: 'pass', detail: 'Valid certificate chain' });
  }

  // Forward secrecy (recommended, not mandated)
  if (!data.forward_secrecy) {
    findings.push({ requirement: 'Best Practice', status: 'warn', detail: 'Forward secrecy not supported (recommended)' });
  } else {
    findings.push({ requirement: 'Best Practice', status: 'pass', detail: 'Forward secrecy enabled' });
  }

  return {
    framework: 'hipaa',
    display_name: 'HIPAA',
    meets_requirements: findings.every(f => f.status !== 'fail'),
    findings,
  };
}

// ─── Public API ─────────────────────────────────────────────────────

export function evaluateCompliance(data: ComplianceInput): ComplianceResult[] {
  return [
    evaluatePCI(data),
    evaluateNIST(data),
    evaluateHIPAA(data),
  ];
}
